ratings/rate_subdir.py

```diff
@@ -0,0 +1,126 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import sys
+sys.path.insert(0, '.')
+
+from absl import flags
+from absl import app
+
+import choix
+import fnmatch
+import numpy as np
+import os
+import re
+import random
+import subprocess
+import math
+from tqdm import tqdm
+import datetime as dt
+from collections import defaultdict
+from ratings import math_ratings
+
+
+FLAGS = flags.FLAGS
+
+PW_REGEX = "PW\[([^]]*)\]"
+PB_REGEX = "PB\[([^]]*)\]"
+RESULT_REGEX = "RE\[([^]]*)\]"
+
+
+def extract_pairwise(files):
+    ids = defaultdict(lambda: len(ids))
+    results = []
+    for _file in tqdm(files):
+        with open(_file) as f:
+            text = f.read()
+        pw = re.search(PW_REGEX, text)
+        pb = re.search(PB_REGEX, text)
+        result = re.search(RESULT_REGEX, text)
+        if not (pw and pb and result):
+            print("Player or result fields not found: ", _file)
+            continue
+
+        pw = pw.group(1)
+        pb = pb.group(1)
+        result = result.group(1)
+
+        if pw == pb:
+            print("Players were the same: ", _file)
+            continue
+        if result.lower().startswith('b'):
+            results.append( (ids[pb], ids[pw]) )
+        if result.lower().startswith('w'):
+            results.append( (ids[pw], ids[pb]) )
+    
+    return ids, results
+
+
+def fancyprint_ratings(ids, ratings, results=None):
+    player_lookup = {v:k for k,v in ids.items()}
+    HEADER = "\n{:<25s}{:>8s}{:>8s}{:>8}{:>7}-{:<8}" 
+    ROW = "{:<25.23s} {:6.0f}  {:6.0f}  {:>6d}  {:>6d}-{:<6d}"
+
+    if not results:
+        for pid, (rating, sigma) in sorted(ratings.items(), 
+                                           key=lambda i: i[1][0], 
+                                           reverse=True):
+            print("{:25s}\t{:5.1f}\t{:5.1f}".format(player_lookup[pid], rating, sigma))
+        return
+
+    wins = {pid : sum([1 for r in results if r[0] == pid]) for pid in ids.values()}
+    losses = {pid : sum([1 for r in results if r[1] == pid]) for pid in ids.values()}
+
+    
+    print("\n{} games played among {} players\n".format(len(results), len(ids)))
+    print(HEADER.format("Name", "Rating", "Error", "Games", "Win", "Loss"))
+    max_r = max(v[0] for v in ratings.values())
+    for pid, (rating, sigma) in sorted(ratings.items(), 
+                                       key=lambda i: i[1][0],
+                                       reverse=True):
+        if rating != max_r:
```

Comment:
I don't see why you wouldn't want to subtract from the max rating too.

if you had ratings 0, 10, 100 now you get -100, -90, 100?

Reply:
The top entry in the table gives you the rating of the highest, and everything else is indexed from there.

So it's like 
```
Stockfish: 3215
challenger1: -20
challenger2: -30
```

The idea is "how far are they from the leader, and where is the leader?"

Reply:
I don't think that's what will happen, because we reset the lowest player to 0 elo you'll always end up with

```
stockfish: 500
challenger1: -X
challenger2: -Y
...
challengerX: -500
```

Stockfish = 3215 is normally anchored from some external benchmark.

Reply:
Yeah but that's fine; if it's like 

Good_model: 2000
crap_test: -2000
or
good_model: 20
decent_test: -20  it still tells us what we want to know.